cpp_exercise/src/lib/ipcmsgqlib.cc

```diff
@@ -0,0 +1,166 @@
+
+   
+
+    int err;
+    remove(_msg_queue_name.c_str());
+        ret = mq_unlink(_msg_queue_name.c_str());
+
+    
+     memset(&_attrs, 0, sizeof _attrs);
+   
+
+    _attrs.mq_maxmsg = _msg_queue_maxnummsg;
+    _attrs.mq_msgsize = _msg_queue_msgsize;
+
+   FileHandler fd(_file_name);
+    
+    std::cout<<"          File Name: "<<_file_name<<std::endl;
+    std::cout<<"          File Size: "<<fd.getSize()<<" bytes"<<std::endl;
+        std::cout<<"          Loading the File. Please Wait..."<<std::endl;
+
+
+    
+
+
+    _msg_queue = mq_open(_msg_queue_name.c_str(), O_WRONLY | O_CREAT , S_IRWXU | S_IRWXG, &_attrs);
+     if(_msg_queue<0){
+         std::cout<<strerror(errno)<<std::endl;
+         throw IPCException("IPCSender ERROR: Cannot open the MSGQueue.");
+
+    }
+   
+ std::cout<<"          MSGQueue is Open: Trying to send the file  "<<std::endl;
+ 
+ fd.openForReading();
+
+        
+
+while (true)
+{
+    std::vector<unsigned char> buffer;
+    buffer = fd.readFile();
+  
+
+    err = mq_send(_msg_queue, (char *)&buffer[0],buffer.size(),_priority);
+      if(buffer.size()==0){
+        break;
+    }
+}
+
+   
+
+	 	 	// check if the queue is empty
+	 	 	while (_check_empty != 0)
+	 	 	{   
+	 	 		mq_getattr(_msg_queue, &_attrs);
+	 	 		_check_empty = _attrs.mq_curmsgs;
+
+	 	 	}
+    
+    
+      
+
+       std::cout<<"          Successfully written to the MSGQueue "<<std::endl;
+       std::cout<<"          MSGQueue is Closed  "<<std::endl;
+
+        std::cout<<"**************************************************************"<<std::endl;
+        std::cout<<"****************************IPCSender: MESSAGE QUEUE PROTOCOL**************************"<<std::endl;
+        std::cout<<"**************************************************************"<<std::endl;
+        std::cout<<"**************************************************************"<<std::endl;
+}
+
+MsgQueueSender::~MsgQueueSender(){
+      mq_close(_msg_queue);
+      
+    
+    mq_unlink(_msg_queue_name.c_str());
+
+}
+
+/***
+ * MsgQueueReceiver Class
+ * 
+ * 
+ * **/
+
+
+MsgQueueReceiver::MsgQueueReceiver(const std::string & filename): _file_name(filename){
+        std::cout<<"**************************************************************"<<std::endl;
+        std::cout<<"****************************IPCReceiver: MESSAGE QUEUE PROTOCOL**************************"<<std::endl;
+        std::cout<<"**************************************************************"<<std::endl;
+        std::cout<<"**************************************************************"<<std::endl;
+
+
+  
+
+        memset(&_attrs, 0, sizeof _attrs); 
+
+
+	 	 	// check if the queue is empty
+	 	 	while (_check_empty == 0)
+	 	 	{ 
+                std::cout<<"          Waiting for the Sender..."<<std::endl;     
+               _msg_queue = mq_open(_msg_queue_name.c_str(), O_RDONLY , &_attrs);  
+	 	 		mq_getattr(_msg_queue, &_attrs);
```

Comment:
I am not sure about the behavior of mq_getattr if the queue is not opened, it may be undeterminated. Maybe you should check that the queue is really opened (_msg_queue >= 0);

Reply:
Fixed